components/org.wso2.carbon.identity.oauth.endpoint/src/main/java/org/wso2/carbon/identity/oauth/endpoint/authz/OAuth2AuthzEndpoint.java

```diff
@@ -2381,6 +2419,32 @@ private String doUserAuthorization(OAuthMessage oAuthMessage, String sessionData
         }
     }
 
+    /**
+     * Validate scopes before consent page
+     *
+     * @param  oAuthMessage oAuthMessage
+     * @param oauth2Params oauth2Params
+     */
+    private void validateScopesBeforeConsent(OAuthMessage oAuthMessage, OAuth2Parameters oauth2Params,
+                                             OAuth2AuthorizeReqDTO authzReqDTO)
+            throws IdentityOAuth2UnauthorizedScopeException, OAuthSystemException {
+
+        try {
+            OAuthAuthzReqMessageContext authzReqMsgCtx = getOAuth2Service().validateScopesBeforeConsent(authzReqDTO);
+            // Add OAuthAuthzReqMessageContext to SessionDataCacheEntry, because we may lose validated scopes if IS
```

Comment:
Here we need to preserve the OAuthAuthzReqMessageContext to preserve backward compatibility as extensions might add information to context that needs to be available when authorizing (issue code, token) the request later.